docs-client/src/components/VariableList/index.tsx

```diff
@@ -56,39 +63,194 @@ export default function({
+          />
+        </TableBody>
+      </Table>
+      <Typography variant="body1" paragraph />
+    </>
+  );
+}
+
+interface State {
+  isEmpty: boolean;
+  isBeans: boolean[];
+  openBeans: boolean[];
+  colSpanLength: number;
+  childStates: { [key: number]: State };
+}
+
+interface OwnProps {
+  indent: number;
+  hasBean: boolean;
+  state?: State;
+  setChildStateInParent?: (index: number, childState: State) => void;
+  childIndex: number;
+}
+
+type FieldInfosProps = OwnProps & Props;
+
+class FieldInfos extends React.Component<FieldInfosProps, State> {
+  private static generateKey(pre: string): string {
+    return `${pre}_${new Date().getTime()}`;
+  }
+
+  private static indentString(indent: number, s: string): string {
+    return `${'\xa0'.repeat(indent)}${s}`;
+  }
+
+  private static printIfNotUnspecified(s: string): string {
+    const lowerCase = s.toLowerCase();
+    if ('unspecified' === lowerCase) {
+      return '-';
+    }
+
+    return lowerCase;
+  }
+
+  constructor(props: FieldInfosProps) {
+    super(props);
+
+    this.setChildStateInParent = this.setChildStateInParent.bind(this);
+
+    if (props.state) {
+      this.state = props.state;
+      return;
+    }
+
+    const fieldInfos = props.variables;
+
+    const isEmpty = fieldInfos.length === 0;
+    let isBeans: boolean[] = [];
+    let openBeans: boolean[] = [];
+    if (!isEmpty) {
+      isBeans = fieldInfos.map(
+        ({ childFieldInfos }) =>
+          !!childFieldInfos && childFieldInfos.length > 0,
+      );
+
+      openBeans = Array(fieldInfos.length).fill(false);
+    }
+
+    let colSpanLength = 4;
+    if (this.props.hasLocation) {
+      colSpanLength += 1;
+    }
+    if (this.props.hasBean) {
+      colSpanLength += 1;
+    }
+
+    const childStates = {};
+    this.state = {
+      isEmpty,
+      isBeans,
+      openBeans,
+      colSpanLength,
+      childStates,
+    };
+  }
+
+  public render() {
+    return (
+      <>
+        {!this.state.isEmpty ? (
+          this.props.variables.map((variable, index) => (
+            <>
+              <TableRow
+                key={FieldInfos.generateKey(variable.name)}
+                onClick={() => this.handleCollapse(index)}
+              >
                 <TableCell>
-                  <code>{variable.name}</code>
+                  <code>
+                    {FieldInfos.indentString(this.props.indent, variable.name)}
+                  </code>
                 </TableCell>
-                {hasLocation && (
-                  <TableCell>
-                    <code>{variable.location}</code>
-                  </TableCell>
-                )}
-                <TableCell>{variable.requirement.toLowerCase()}</TableCell>
+                {this.props.hasLocation &&
+                  variable.location && (
+                    <TableCell>
+                      <code>
+                        {FieldInfos.printIfNotUnspecified(
+                          variable.location.toLowerCase(),
```

Comment:
No need to call `toLowerCase()` because `printIfNotSpecified()` does it for the caller.